src/ui/public/visualize/loader/embedded_visualize_handler.js

```diff
@@ -26,58 +30,141 @@ const RENDER_COMPLETE_EVENT = 'render_complete';
+
+    return this._visualize.fetch(this._params, forceFetch);
+  };
+
+  _render = (visData) => {
+    return visualizationLoader(this._element, this._vis, visData, this._uiState, { listenOnChange: false }).then(() => {
+      if (!this._loaded) {
+        this._loaded = true;
+        this._fetchAndRender();
+      }
+    });
+  };
+
+  _fetchAndRender = _.debounce((forceFetch) => {
+    return this._fetch(forceFetch).then(this._render);
+  }, 100);
+
   /**
    * Update properties of the embedded visualization. This method does not allow
    * updating all initial parameters, but only a subset of the ones allowed
    * in {@link VisualizeLoaderParams}.
    *
    * @param {Object} [params={}] The parameters that should be updated.
    * @property {Object} [timeRange] A new time range for this visualization.
+   * @property {Object} [filters] New filters for this visualization.
+   * @property {Object} [query] A new query for this visualization.
    * @property {Object} [dataAttrs] An object of data attributes to modify. The
    *    key will be the name of the data attribute and the value the value that
    *    attribute will get. Use null to remove a specific data attribute from the visualization.
    */
   update(params = {}) {
-    this._scope.$evalAsync(() => {
-      if (params.hasOwnProperty('timeRange')) {
-        this._scope.timeRange = params.timeRange;
-      }
-      if (params.hasOwnProperty('filters')) {
-        this._scope.filters = params.filters;
-      }
-      if (params.hasOwnProperty('query')) {
-        this._scope.query = params.query;
-      }
+    // Apply data- attributes to the element if specified
+    if (params.dataAttrs) {
+      Object.keys(params.dataAttrs).forEach(key => {
+        if (params.dataAttrs[key] === null) {
+          this._element.removeAttribute(`data-${key}`);
+          return;
+        }
 
-      // Apply data- attributes to the element if specified
-      if (params.dataAttrs) {
-        Object.keys(params.dataAttrs).forEach(key => {
-          this._element.attr(`data-${key}`, params.dataAttrs[key]);
-        });
-      }
-    });
+        this._element.setAttribute(`data-${key}`, params.dataAttrs[key]);
+      });
+    }
+
+    let fetchRequired = false;
+    if (params.hasOwnProperty('timeRange')) {
+      fetchRequired = true;
+      this._params.timeRange = params.timeRange;
+    }
+    if (params.hasOwnProperty('filters')) {
+      fetchRequired = true;
+      this._params.filters = params.filters;
+    }
+    if (params.hasOwnProperty('query')) {
+      fetchRequired = true;
+      this._params.query = params.query;
+    }
+
+    if (fetchRequired) {
+      this._fetchAndRender();
+    }
   }
 
   /**
    * Destroy the underlying Angular scope of the visualization. This should be
    * called whenever you remove the visualization.
    */
   destroy() {
-    this._scope.$destroy();
+    this._destroyed = true;
+    this._vis.removeListener('reload', this._reloadVis);
```

Comment:
I think there might be a tiny chance, that we're actually currently have triggered a `fetchAndRender` that's still in debounce while we're destroying. I think in that case we should cancel the debounce. It seems the method returned from the `debounce` call above should have a property `cancel` so that we can call `this._fetchAndRender.cancel()` here.